Lib9c/Model/State/ShardedShopState.cs

```diff
@@ -0,0 +1,112 @@
+{
+    public class ShardedShopState : State
+    {
+        public static Address DeriveAddress(ItemSubType itemSubType, Guid productId)
+        {
+            switch (itemSubType)
+            {
+                case ItemSubType.Weapon:
+                case ItemSubType.Armor:
+                case ItemSubType.Belt:
+                case ItemSubType.Necklace:
+                case ItemSubType.Ring:
+                case ItemSubType.Food:
+                    string nonce = productId.ToString().Substring(0, 1);
+                    return DeriveAddress(itemSubType, nonce);
+                case ItemSubType.FullCostume:
+                case ItemSubType.HairCostume:
+                case ItemSubType.EarCostume:
+                case ItemSubType.EyeCostume:
+                case ItemSubType.TailCostume:
+                case ItemSubType.Title:
+                    return Addresses.Shop.Derive($"{itemSubType}");
+                default:
+                    throw new InvalidItemTypeException($"Unsupported ItemType: {itemSubType}");
+            }
+        }
+
+        public static readonly IReadOnlyList<string> AddressKeys = new List<string>
+        {
+            "0",
+            "1",
+            "2",
+            "3",
+            "4",
+            "5",
+            "6",
+            "7",
+            "8",
+            "9",
+            "a",
+            "b",
+            "c",
+            "d",
+            "e",
+            "f",
+        };
+
+        public static Address DeriveAddress(ItemSubType itemSubType, string nonce)
+        {
+            switch (itemSubType)
+            {
+                case ItemSubType.Weapon:
+                case ItemSubType.Armor:
+                case ItemSubType.Belt:
+                case ItemSubType.Necklace:
+                case ItemSubType.Ring:
+                case ItemSubType.Food:
+                    return Addresses.Shop.Derive($"{itemSubType}-{nonce}");
+                case ItemSubType.FullCostume:
+                case ItemSubType.HairCostume:
+                case ItemSubType.EarCostume:
+                case ItemSubType.EyeCostume:
+                case ItemSubType.TailCostume:
+                case ItemSubType.Title:
+                    return Addresses.Shop.Derive($"{itemSubType}");
+                default:
+                    throw new InvalidItemTypeException($"Unsupported ItemType: {itemSubType}");
+            }
+        }
+
+
+        public readonly Dictionary<Guid, ShopItem> Products = new Dictionary<Guid, ShopItem>();
+
+        public ShardedShopState(Address address) : base(address)
+        {
+        }
+
+        public ShardedShopState(Dictionary serialized) : base(serialized)
+        {
+            Products = serialized[ProductsKey]
+                .ToList(s => new ShopItem((Dictionary) s))
+                .ToDictionary(s => s.ProductId, s => s);
+        }
+
+        public ShardedShopState(IValue iValue) : this((Dictionary)iValue)
+        {
+        }
```

Comment:
IMHO this constructor seems useless because other constructor takes `Dictionary` as argument and the type system will check it.

```
        public ShardedShopState(Dictionary serialized) : base(serialized)
        {
            Products = serialized[ProductsKey]
                .ToList(s => new ShopItem((Dictionary) s))
                .ToDictionary(s => s.ProductId, s => s);
        }
```

Reply:
Since this constructor is only being used for testing, I will delete it.